e2e/test/scenarios/native/native-database-source.cy.spec.js

```diff
@@ -0,0 +1,247 @@
+        .should("have.text", "Sample Database")
+        .click();
+
+      cy.log("Pick the same database again");
+      selectDatabase("Sample Database");
+      cy.get("@persistDatabase").should("be.null");
+    });
+
+    it("selecting a database in native editor for model actions should not persist the database", () => {
+      [SAMPLE_DB_ID, PG_DB_ID].forEach(enableModelActionsForDatabase);
+
+      startNewAction();
+      assertNoDatabaseSelected();
+
+      selectDatabase("Sample Database");
+      cy.get("@persistDatabase").should("be.null");
+
+      startNativeModel();
+      assertNoDatabaseSelected();
+      cy.log(
+        "Persisting a database for a native model should not affect actions",
+      );
+      selectDatabase("QA Postgres12");
+      cy.wait("@persistDatabase");
+
+      startNewAction();
+      assertNoDatabaseSelected();
+    });
+
+    describe("permissions", () => {
+      it("users with 'No self-service' data permissions should be able to choose only the databases they can query against", () => {
+        cy.signIn("nodata");
+
+        startNativeQuestion();
+        cy.wait("@persistDatabase");
+        cy.findByTestId("selected-database")
+          .should("have.text", "QA Postgres12")
+          .click();
+
+        cy.get(POPOVER_ELEMENT).should("not.exist");
+
+        cy.signOut();
+        cy.signInAsAdmin();
+
+        addPostgresDatabase(additionalPG);
+
+        cy.signIn("nodata");
+        startNativeQuestion();
+
+        cy.findByTestId("selected-database")
+          .should("have.text", "QA Postgres12")
+          .click();
+
+        popover()
+          .should("contain", "QA Postgres12")
+          .and("contain", "New Database");
+      });
+    });
+
+    it.skip("users with no native write permissions should be able to choose only the databases they can query against (metabase#39053)", () => {
+      cy.signIn("nosql");
+
+      startNativeQuestion();
+      cy.wait("@persistDatabase");
+      cy.findByTestId("selected-database")
+        .should("have.text", "QA Postgres12")
+        .click();
+
+      cy.get(POPOVER_ELEMENT).should("not.exist");
+    });
+  },
+);
+
+describe("mongo as the default database", { tags: "@mongo" }, () => {
+  beforeEach(() => {
+    restore("mongo-4");
+    cy.signInAsAdmin();
+  });
+
+  it("should persist Mongo database, but not its selected table", () => {
+    startNativeQuestion();
+    assertNoDatabaseSelected();
+
+    selectDatabase(mongoName);
+    cy.findByTestId("native-query-top-bar")
+      .findByText("Select a table")
+      .click();
+    popover().findByText("Reviews").click();
+    cy.findByTestId("native-query-top-bar").should(
+      "not.contain",
+      "Select a table",
+    );
+
+    startNativeQuestion();
+
+    assertSelectedDatabase(mongoName);
+    cy.findByTestId("native-query-top-bar").should("contain", "Select a table");
+  });
+});
+
+function startNativeQuestion() {
+  cy.visit("/");
```

Comment:
This is expensive. Can we use SPA navigation instead?

```suggestion
  cy.findByTestId("main-logo-link").click();
```

----

The same comment applies to other `cy.visit` calls in this test suite.

Reply:
It is expensive, but it was the only reliable approach that gave me confidence this thing works.
I can explore some other options.